TankDriver.App/Models/BulletSpaceModel.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using TankDriver.Logic;
+
+namespace TankDriver.Models
+{
+	internal class BulletSpaceModel: IModel
+	{
+		private BulletSpace _bulletSpace;
+		private TextureStorage _textureStorage;
+
+		public BulletSpaceModel (BulletSpace bulletSpace)
+		{
+			_bulletSpace = bulletSpace;
+		}
+
+		public void LoadBulletTextures(IModel bulletModel)
+		{
+			bulletModel.LoadTextures (_textureStorage);
+		}
+
+		#region IModel implementation
+
+		public IUnit GetUnit ()
+		{
+			return _bulletSpace;
+		}
+
+		public void LoadTextures (TextureStorage textureStorage)
+		{
+			_textureStorage = textureStorage;
+		}
+
+		public void Render (SpriteBatch spriteBatch)
+		{
+			foreach (var bullet in _bulletSpace.Bullets) {
+				bullet.GetModel ().Render (spriteBatch);
+			}
+		}
+
+		#endregion
+	}
+}
+
```

Comment:
I don't like this class either. I was just implementing by analogy trying to fit my code into this `IUnit-IModel` system. To be honest I don't understand why are we trying to separate `IUnit`-s and `IModel`-s physically making them completely different objects. For me IUnit and IModel are just two different interfaces for the same object which incapsulates the behaviour of an entity.

Anyway, I simply didn't know that you was planning to use that `IUnit-IModel` system only for some sort of "atomical" entities. I will get rid of that `BulletSpaceModel` then.

> Yes, we do have a need to store a list of the models somewhere, but let's maybe just store the list in the Game class itself?

So you suggest to store and process like *everything* in a single object?

I had the following domain field decomposition in my head. Please let me know if you see any problems with it.

We have the following entities:

- `Tank` -- which is the tank;
- `Bullet` -- a projectile that the `Tank` can shoot;
- `BulletSpace` -- a space where the `Tank` can shoot its `Bullet`s into;
- `Enemy` -- something that the `Tank` can take down with its
  `Bullet`s, maybe it's gonna be called differently, doesn't matter;

On the top level (in the `Game` object to be precise) we have only
`Tank`, `Enemy` and `BulletSpace` (definitely there will be more in
the top level in the future, I'm talking about right now). We
periodically call `Update()` and `Render()` only for those entities.

Ok, the `Tank` wants to shoot a `Bullet`. Let's just ask the `Tank` to shoot
it into the space (see https://github.com/ForNeVeR/TankDriver/pull/20/files#diff-b5292f9a00d8778577c26716c5f4936cR113):

```
Tank.ShootInto(BulletSpace)
```

Alright, looks pretty straightforward and intuitive. At least for me. How is that `ShootInto` method is implemented? (see https://github.com/ForNeVeR/TankDriver/pull/20/files#diff-28f8511e517e79be8f0939b939655cdeR141)

```
class Tank:
    field TurretPosition
    field TurretHeading

    method ShootInto(BulletSpace):
        BulletSpace.AddBullet(TurretPosition, TurretHeading)
```

Pretty straightforward as well. And I really like how that `Tank-BulletSpace` system feels. They are two objects that are simply communicating with each other. In that system they don't even have to expose their guts (with public properties and stuff) to perform a shot. They just tell each other enough information to do that.

With that logic I'd implement the `Tank` itself as a composition of `Body` and `Turret`:

```
class Tank:
    field Turret
    field Body

    method ShootInto(BulletSpace):
        turret.ShootInto(BulletSpace)
```

Because the shot is actually performed by the `Turret` which is the part of the `Tank`. But that's a completely different story.

Ok, now we want to be able to hit the `Enemy` with a `Bullet` in the `BulletSpace`.

```
BulletSpace.hit(Enemy)
```

How I see that implemented:

```
class Enemy:
    property Shape

    method HitAt(Position):
        if Shape.Contains(Position) then
            // die

class Bullet:
    property Position

    method Hit(Enemy):
        Enemy.HitAt(Position)

class BulletSpace:
    property Bullets

    method Hit(Enemy):
        Foreach Bullet In Bullets Do
            Bullet.Hit(Enemy)
```

And we don't actually have to pass the `Enemy` by itself. We can pass it by a small interface like:

```
interface Hitable:
    method HitAt(Position)
```

So a `Bullet` doesn't know about `Enemy`. Which is actually better, because `Bullet` doesn't have enemies. `Bullet` doesn't know good or bad. It just hits objects. `Hitable` objects.

We can also abstract out `BulletSpace` and create interface `Space`. Why just have one space if we can have many? It enables us with so many interesting things!

For example, let's say we have two tanks who fight each other: `RedTank` and `BlueTank`. Each tank shoots into its own space:

```
RedTank.ShootInto(RedBulletSpace)
BlueTank.ShootInto(BlueBulletSpace)
```

and each tank is hit by the bullets of the opposite tank's space!

```
RedBulletSpace.Hit(BlueTank)
BlueBulletSpace.Hit(RedTank)
```

And it's so easy and straightforward to describe the behaviour of entities such way! That is where I was heading when I was implementing the shooting feature.

Maybe I'm just crazy I dunno. Let me know what you think about all of that.